File: packages/py/core/core/algorithms/get_implementations.py
from typing import List, Optional

from ..types import InterfaceImplementations, Uri, UriRedirect
from . import apply_redirects


def get_implementations(
    wrapper_interface_uri: Uri,
    interfaces: List[InterfaceImplementations],
    redirects: Optional[List[UriRedirect]] = None,
) -> List[Uri]:
    if redirects is None:
        redirects = []

    def add_unique_result(uri: Uri, result: List[Uri]):
        if not uri in result:
            result.append(uri)

    def add_all_implementations_from_implementations_array(
        implementations_array: List[InterfaceImplementations],
        api_interface_uri: Uri,
    ) -> List[Uri]:
        result: List[Uri] = []
        for interface_implementations in implementations_array:
            fully_resolved_uri = (
                apply_redirects(interface_implementations.interface, redirects)
                if redirects
                else interface_implementations.interface
            )

            if fully_resolved_uri == api_interface_uri:
                for implementation in interface_implementations.implementations:
                    add_unique_result(implementation, result)
        return result

    final_uri = wrapper_interface_uri
    if redirects:
        final_uri = apply_redirects(wrapper_interface_uri, redirects)

    result = add_all_implementations_from_implementations_array(interfaces, final_uri)

    return result

File: packages/js/core-client/src/algorithms/getImplementations.ts
import {
  Uri,
  InterfaceImplementations,
  CoreClient,
  WrapError,
  WrapErrorCode,
} from "../types";
import { IUriResolutionContext } from "../uri-resolution";
import { applyResolution } from "./applyResolution";

import { Result, ResultErr, ResultOk } from "@polywrap/result";

export const getImplementations = async (
  wrapperInterfaceUri: Uri,
  interfaces: readonly InterfaceImplementations[],
  client?: CoreClient,
  resolutionContext?: IUriResolutionContext
): Promise<Result<Uri[], WrapError>> => {
  const result: Uri[] = [];

  const addUniqueResult = (uri: Uri) => {
    // If the URI hasn't been added already
    if (result.findIndex((i) => Uri.equals(i, uri)) === -1) {
      result.push(uri);
    }
  };

  const addAllImplementationsFromImplementationsArray = async (
    implementationsArray: readonly InterfaceImplementations[],
    wrapperInterfaceUri: Uri
  ): Promise<Result<undefined, WrapError>> => {
    for (const interfaceImplementations of implementationsArray) {
      let fullyResolvedUri: Uri;
      if (client) {
        const redirectsResult = await applyResolution(
          interfaceImplementations.interface,
          client,
          resolutionContext
        );
        if (!redirectsResult.ok) {
          const error = new WrapError("Failed to resolve redirects", {
            uri: interfaceImplementations.interface.uri,
            code: WrapErrorCode.CLIENT_GET_IMPLEMENTATIONS_ERROR,
            cause: redirectsResult.error,
          });
          return ResultErr(error);
        }
        fullyResolvedUri = redirectsResult.value;
      } else {
        fullyResolvedUri = interfaceImplementations.interface;
      }

      if (Uri.equals(fullyResolvedUri, wrapperInterfaceUri)) {
        for (const implementation of interfaceImplementations.implementations) {
          addUniqueResult(implementation);
        }
      }
    }
    return ResultOk(undefined);
  };

  let finalUri = wrapperInterfaceUri;

  if (client) {
    const redirectsResult = await applyResolution(
      wrapperInterfaceUri,
      client,
      resolutionContext
    );
    if (!redirectsResult.ok) {
      const error = new WrapError("Failed to resolve redirects", {
        uri: wrapperInterfaceUri.uri,
        code: WrapErrorCode.CLIENT_GET_IMPLEMENTATIONS_ERROR,
        cause: redirectsResult.error,
      });
      return ResultErr(error);
    }
    finalUri = redirectsResult.value;
  }

  const addAllImp = await addAllImplementationsFromImplementationsArray(
    interfaces,
    finalUri
  );

  return addAllImp.ok ? ResultOk(result) : addAllImp;
};
